### PR TITLE
ARROW-6874: [Python] Fix memory leak when converting to Pandas object data

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
@wesm @jorisvandenbossche would you like to review?

**Comment by wesm** [comment]
Looking.

File: cpp/src/arrow/python/arrow_to_pandas.cc [hunk 1]
```diff
@@ -347,8 +346,17 @@ class PandasBlock {
     } else {
       block_dims[0] = num_rows_;
     }
-    RETURN_NOT_OK(PyArray_NewFromPool(ndim, block_dims, npy_type,
-                                      /*arrow_type=*/nullptr, options_.pool, &block_arr));
+    PyArray_Descr* descr = internal::GetSafeNumPyDtype(npy_type);
+    if (PyDataType_REFCHK(descr)) {
+      // ARROW-6876: if the array has refcounted items, let Numpy
+      // own the array memory so as to decref elements on array destruction
+      block_arr = PyArray_SimpleNewFromDescr(ndim, block_dims, descr);
```
Comment (wesm) [inline cpp/src/arrow/python/arrow_to_pandas.cc]: Another option would be to have a buffer that decrefs the pointers in its dtor, but that seems like a bit too much work here

**Review by wesm — APPROVED** [review]
+1

File: cpp/src/arrow/python/arrow_to_pandas.cc [hunk 2]
```diff
@@ -1824,8 +1832,20 @@ class ArrowDeserializer {
...
-    RETURN_NOT_OK(
-        PyArray_NewFromPool(1, dims, type, data_->type().get(), options_.pool, &result_));
+    PyArray_Descr* descr = internal::GetSafeNumPyDtype(type);
+    if (descr == nullptr) {
+      RETURN_IF_PYERROR();
+    }
+    if (PyDataType_REFCHK(descr)) {
+      // ARROW-6876: if the array has refcounted items, let Numpy
+      // own the array memory so as to decref elements on array destruction
+      set_numpy_metadata(type, data_->type().get(), descr);
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/python/arrow_to_pandas.cc]: Hmm, you're right. Thought it's harmless as well.

**Pull request closed without merging**